### DISCUSSION TITLE
auto add prefix

### DISCUSSION THREAD

**Discussion by IRediTOTO** [tailwindlabs/tailwindcss#2598] [Ideas]
Hi, I love Tailwind 💯 
**Is there any tool can help us add prefix to class name Tw?**
I just add a prefix tw- to use with a webpack plugin (it need to find classname by regex)
I tried to search in my html code and there are >800 class attr :D that mean i must add thousand of tw- to class.

**Comment by simonswiss** [comment]
Hello @IRediTOTO 👋

If I understand your question correctly, the [prefix option](https://tailwindcss.com/docs/configuration#prefix) in the Tailwind config file is exactly what you're looking for!

In your config file, define a string you want to prefix all Tailwind generated classes with:

```js
// tailwind.config.js
module.exports = {
  theme: {
    // ...
  },
  prefix: 'tw-',
}
```

This will prefix all Tailwind classes with `tw-`. 

`bg-red-100` will become `tw-bg-red-100`, etc.

Hope this helps! 👍

**Reply by IRediTOTO** [reply]
Yes, but in js file, which I type thousand of tw class without tw- :D That will be crazy.
So I created this func to do that, hope it help someone
```
const fs=require("fs")

function changeForm() {
    const data=fs.readFileSync("./pages/index.js")
    let list_class = data.match(/(?<=className=")(.*)(?=")/g);
    list_class = list_class.map((classes) => {
       return classes.split(" ");
    });
 
    list_class = [].concat.apply([], list_class);
    list_class=list_class.filter((el)=>el!=="")
    list_class = list_class.map((classes) => {
       if (classes.includes(":")) {
          return classes.split(":")[1];
       } else {
          return classes;
       }
    });
 
    list_class = [].concat.apply([], list_class);
    list_class=list_class.filter((el)=>el!=="")
    let old_class = list_class;
    list_class = list_class.map((classes) => {
       if (!classes.includes("tw-")) {
          return "tw-" + classes;
       } else {
          return classes;
       }
    });
 
    let return_data = data.split(" ");
    for (i = 0; i < return_data.length - 1; i++) {
       for (j = 0; j < old_class.length - 1; j++) {
 
          if(return_data[i].trim().includes(old_class[j])&&return_data[i]!==""){
             return_data[i]=return_data[i].replace(old_class[j],list_class[j])
             break;
          }
          
          console.log("list_class",return_data[i])
       }
    }
    console.log("old_class",old_class)
    console.log("list_class",list_class)
    return_data=return_data.join(" ")
 
    return return_data;
 }
let a=`<div className=" flex flex-row hover:items-center  justify-between ">
<h3 className=" text-gray-700 font-semibold text-lg">Book like a boss</h3>
<p>500$</p>
</div>`
console.log(changeForm(a))
```

**Reply by FathurrahmanRifqi** [reply]
cool, you can try this too 
https://github.vue.tailwind-prefix.cbass.dev/

**Reply by johnantoni** [reply]
excellent! thanks for putting this together

**Reply by danielduckworth** [reply]
> cool, you can try this too https://github.vue.tailwind-prefix.cbass.dev/

Nice one! Exactly what I was looking for. This would be a great VS code extension.

**Reply by PositiveUser** [reply]
> cool, you can try this too https://github.vue.tailwind-prefix.cbass.dev/

Thank you very much. Unfortunately, this did not work for classes with arbitrary value.

**Reply by Folyd** [reply]
> cool, you can try this too https://github.vue.tailwind-prefix.cbass.dev/

Great tool! Can you also support arbitrary values? Thanks.

**Reply by Parth909** [reply]
> cool, you can try this too https://github.vue.tailwind-prefix.cbass.dev/

Great tool ! Thank you for making it.

**Reply by Parth909** [reply]
> > cool, you can try this too https://github.vue.tailwind-prefix.cbass.dev/
> 
> Great tool! Can you also support arbitrary values? Thanks.

You can do `Search (Ctrl + F)` in your editor -> turn on *Regex Option* -> use this Regex `[\w]*-\[[\w|.]*\]`. This will search for JIT styles (arbitrary values)

**Reply by nischalj10** [reply]
Lifesaver!

**Reply by SySagar** [reply]
hey , as per my knowlege the prefix doesn't get added automatically. So is there a way to auto add it?

**Comment by simonswiss** [comment]
If you define a `prefix: 'tw-'` in the config file, and use the Tailwind Intellisense vscode extension, you'll have autocomplete suggestion for each class with the prefix applied 👍

![image](https://user-images.githubusercontent.com/485747/96537438-01616100-12e2-11eb-8fb5-13243387545d.png)

But like you said, there are probably situations where this is not helpful (classes are already written in the HTML/JSX), and in that case you need to come up with a clever approach like your `changeForm` transform function.

**Reply by m-ostadi** [reply]
You can use a regex like [the one I wrote in this gist](https://gist.github.com/MOstadi/489408713b60561a2f3e92acb1c0c191) to find and replace all existing class with prefixed version.

**Reply by Parth909** [reply]
> You can use a regex like [the one I wrote in this gist](https://gist.github.com/MOstadi/489408713b60561a2f3e92acb1c0c191) to find and replace all existing class with prefixed version.

This doesn't exist anymore

**Reply by IRediTOTO** [reply]
[prefix](https://github.vue.tailwind-prefix.cbass.dev/)
You want this?

**Reply by Parth909** [reply]
> [prefix](https://github.vue.tailwind-prefix.cbass.dev/) You want this?

Yep this is what I'm using now

**Reply by m-ostadi** [reply]
> > You can use a regex like [the one I wrote in this gist](https://gist.github.com/MOstadi/489408713b60561a2f3e92acb1c0c191) to find and replace all existing class with prefixed version.
> 
> This doesn't exist anymore

Sorry i changed my username on github, this is the correct link:

https://gist.github.com/m-ostadi/489408713b60561a2f3e92acb1c0c191

**Comment by restieauxbro** [comment]
When I wanted to add a prefix inav- to all of my classes in the markup, I used this node script.

```javascript
const fs = require('fs');
const path = require('path');
const glob = require("glob");

const rootDir = "path/to/your/repo";
const classNameRegex = /className=["']([\w\s-]+)["']/g;

glob("**/*.{js,jsx,tsx}", {cwd: rootDir}, (er, files) => {
  files.forEach((file) => {
    const filePath = path.join(rootDir, file);
    fs.readFile(filePath, 'utf-8', (err, data) => {
      if (err) {
        console.error(err);
        return;
      }
      const result = data.replace(classNameRegex, (match, p1) => {
        const classNames = p1.split(" ");
        const prefixedClassNames = classNames.map((className) => {
          if (className.startsWith("inav-")) {
            return className;
          } else {
            return `inav-${className}`;
          }
        });
        return `className="${prefixedClassNames.join(" ")}"`;
      });
      fs.writeFile(filePath, result, 'utf-8', (err) => {
        if (err) console.error(err);
      });
    });
  });
});
```

This will allow you to change all of your already written classes to give them the tailwind prefix you want.

**Reply by mnzsss** [reply]
Great tool ! Thank you for making it.

**Reply by hereisfahad** [reply]
@restieauxbro Thanks for this solution 🙌 but this does not work for conditional classes. Is it possible to update all the strings within the JSX or, even better, update the classes by checking if they are valid Tailwind CSS classes, like performing some form of parsing on the JSX?

**Reply by juliushuck** [reply]
Worked for me with the following glob version

"glob": "^7.2.3",

**Comment by pedroapfilho** [comment]
It would be nice to add this to the prettier plugin, or add it to a library that we can use to auto-add the prefixes

**Comment by unional** [comment]
The `prefix` config doesn't work when it involves shared libraries.

I have a MFEs project which each MFEs has their own prefix. The shared library can't define a prefix that works for every MFEs.

That's why a bundle-time plugin with `webpack`, `rollup`, or even `ttypescript` would help.

Also, another thing is about tree-shaking, which can only take effect during bundling.

I don't know if there is a way to make it work in Tailwind config:

```
module.exports = {
  // this doesn't work when `some-shared-library` not using the same prefix
  prefix: 'abc-',
  content: [
    './src/**/*.tsx',
    // this will get all css classes used in the library, not tree shaked.
    './node_modules/some-shared-library/dist/**/*.js'
  ]
 }
 ```

**Reply by tejas-1206** [reply]
@unional Found any solution? I'm in a similar situation as yours.

**Reply by unional** [reply]
Not really, we include the full css of the library as a workaround.

**Reply by tejas-1206** [reply]
> Not really, we include the full css of the library as a workaround.

As in the full CSS file generated by tailwind ?

**Comment by nabeel-shakeel** [comment]
You can use this tool to covert your classes into prefixed ones. Covers all the cases of latest tailwind
https://tailwind-prefix-tool.netlify.app/

**Reply by kyr0** [reply]
Thank you, great tool!

**Reply by pedroapfilho** [reply]
It would be great to have it as a prettier plugin!

**Reply by otherperspectives** [reply]
Is this still working? I have tried using it and it did not work as expected..

**Reply by pedroapfilho** [reply]
It doesn't have some new properties, but it still works

**Comment by kaminskypavel** [comment]
it would be nice if can have a codemod for that

**Reply by pedroapfilho** [reply]
codemod, or a prettier autofix

**Reply by spa5k** [reply]
A build time codemod will be very good and necessary.

**Comment by aryomuzakki** [comment]
Some of the tools in this discussion or other just don't active anymore or not covering some cases, here i just made one that work for v3.4.4.

https://tw-prefixer.vercel.app/

This works for negative classes, arbitrary classes, variants/modifiers, and !important classes in html or jsx.
Classnames are crawled from tailwind docs with a few addition. 

Still have some flaws cause tailwind is just that great and flexible in providing more arbitrary values / selectors, and modifiers.
Your feedback will be very appreciated

**Reply by kyr0** [reply]
Could you maybe offer the algo as an NPM package? A simple function being exported allowing to pass in code just as with the UI and having the post-processed text returned would allow for a postcss plugin to be written :)

**Reply by SySagar** [reply]
@aryomuzakki  why not package this into a npm module and then we can use this my just runnning a pre script before building tailwind classess?

**Reply by nehallsharma** [reply]
This doesn't work for dynamic classes

**Comment by GlaceDream** [comment]
I wrote a script to add prefix to all my classes of my React TypeScript project but actually it was useless because I used NextUi which does not support tailwind prefix.  So, to not waste my work and my time, I prefer to share my script. I hope it will help you. The script is here if you need: 

https://github.com/JeremieLevain/addTailwindPrefix/blob/f4bb0675b898509370b34102a1918b491ac1e599/addTailwindPrefix.js

I used the package "classnames" imported as "cn", so my RegExp takes it in account.

**Comment by Maks-T** [comment]
https://www.npmjs.com/package/tailwind-prefix-migrator

there is a utility that adds a prefix and isolates it in a separate layer. It also works with
